src/vcpkg/binarycaching.cpp

```diff
@@ -1449,6 +1490,34 @@ namespace vcpkg
         return results;
     }
 
+    void BinaryCache::push_thread_main()
+    {
+        while (true)
+        {
+            std::unique_lock<std::mutex> lock(actions_to_push_mutex);
+            actions_to_push_notifier.wait(lock, [this]() { return !actions_to_push.empty() || end_push_thread; });
+            if (actions_to_push.empty())
+            {
+                if (end_push_thread) break;
+                continue;
+            }
+
+            auto entry = std::move(actions_to_push.front());
+            actions_to_push.pop();
+            lock.unlock(); // we don't touch actions_to_push anymore
+
+            for (auto&& provider : m_providers)
+            {
+                provider->push_success(entry.info, entry.packages_dir, stdout_sink);
```

Comment:
We can't use stdout_sink for this -- it would mean tearing between threads. This requires a novel `MessageSink` that's a SPSC queue of lines. Roughly:

```cpp
struct BGMessageSink : MessageSink
{
    std::mutex m_lock;
    // guarded by m_lock
    std::vector<std::pair<Color, string>> m_published;
    // unguarded, buffers messages until newline is reached
    std::vector<std::pair<Color, string>> m_unpublished;

    // must be called from producer
    void print(Color c, string s) override
    {
        auto pos = s.find_last('\n');
        if (pos != string::npos)
        {
            std::lock_guard<std::mutex> lk(m_lock);
            m_published.insert(m_published.end(),
                               std::make_move_iterator(m_unpublished.begin()),
                               std::make_move_iterator(m_unpublished.end()));
            m_published.emplace_back(c, s.substr(0, pos + 1));
        }
        m_unpublished.clear();
        if (s.size() > pos + 1)
        {
            m_unpublished.emplace_back(c, s.substr(pos + 1));
        }
    }

    // must be called from consumer (synchronizer of out)
    void print_published(MessageSink& out)
    {
        std::lock_guard<std::mutex> lk(m_lock);
        for (auto&& m : m_published)
        {
            out.print(m.first, m.second);
        }
        m_published.clear();
    }

    // must be called from producer / after producer finishes
    void flushall()
    {
        std::lock_guard<std::mutex> lk(m_lock);
        m_published.insert(m_published.end(),
                           std::make_move_iterator(m_unpublished.begin()),
                           std::make_move_iterator(m_unpublished.end()));
    }
};
```

Then, the outer owner of binary caching can call `.print_published(stdout_sink)` between packages to emit every complete line produced by the binary caching machinery. This at least limits tearing to line-based.

It would also be reasonable to not flush on lines and instead require manual flushing from the consumer thread; this would entirely prevent tearing and ensure we emit all the messages associated with a specific binary upload all at once.

Before destroying this sink, the owner should call `.flushall()` and `.print_published(stdout_sink)` a final time to make sure everything was properly printed.

There's also a _slightly_ higher perf option since this is a SPSC queue -- there could be a third `m_post_published` vector which is swapped with `m_published` under the lock. This would get the actual prints out from under the lock which means any blocking on stdout would not prevent more binary packages from being uploaded in the background. The `swap` would ensure that at least the vector's memory gets reused without constant reallocation -- though since most of the allocations here are the individual strings I'm not sure it makes a difference.